src/PowerShellEditorServices/Session/PowerShellContext.cs

```diff
@@ -823,44 +823,16 @@ public Task ExecuteCommand(PSCommand psCommand)
 
         internal static TResult ExecuteScriptAndGetItem<TResult>(string scriptToExecute, Runspace runspace, TResult defaultValue = default(TResult))
         {
-            Pipeline pipeline = null;
-
-            try
+            using (var pwsh = PowerShell.Create())
             {
-                if (runspace.RunspaceAvailability == RunspaceAvailability.AvailableForNestedCommand)
-                {
-                    pipeline = runspace.CreateNestedPipeline(scriptToExecute, false);
-                }
-                else
-                {
-                    pipeline = runspace.CreatePipeline(scriptToExecute, false);
-                }
-
-                Collection<PSObject> results = pipeline.Invoke();
+                pwsh.Runspace = runspace;
+                Collection<TResult> results = pwsh.AddScript(scriptToExecute).Invoke<TResult>();
 
                 if (results.Count == 0 || results.FirstOrDefault() == null)
                 {
                     return defaultValue;
                 }
-
-                if (typeof(TResult) != typeof(PSObject))
-                {
-                    return results
-                            .Select(pso => pso.BaseObject)
-                            .OfType<TResult>()
-                            .FirstOrDefault();
-                }
-                else
-                {
-                    return
-                        results
-                            .OfType<TResult>()
-                            .FirstOrDefault();
-                }
-            }
-            finally
-            {
-                pipeline.Dispose();
+                return results.FirstOrDefault();
```

Comment:
Does this make the check above it redundant?

Reply:
no because you can pass in an alternative default value if you'd like.

Reply:
Oh right. Well the two calls to `FirstOrDefault` still feel a bit redundant, and the first call is only relevant to reference types anyway.

So we could achieve the same with:
```csharp
if (results.Count == 0)
{
    return defaultValue;
}

return results.FirstOrDefault() ?? defaultValue;
```

Just a thought though

Reply:
Technically the final thing could just be `results.First()` since having no results there is either impossible or a very strange bug...

Reply:
How about

```csharp
results.DefaultIfEmpty(defaultValue).FirstOrDefault();
```

Reply:
Oooh, yes! I thought that was a thing, but never used it before...

Reply:
fixed :)